pkg/api_gateway/api_gateway.go

```diff
@@ -0,0 +1,222 @@
+package apigateway
+
+import (
+	"encoding/hex"
+	"errors"
+	"sync"
+
+	"github.com/centrifuge/go-substrate-rpc-client/v4/types"
+	"github.com/rs/zerolog/log"
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	"github.com/threefoldtech/zos/pkg"
+)
+
+type apiGateway struct {
+	sub      *substrate.Substrate
+	mu       sync.Mutex
+	identity substrate.Identity
+}
+
+func NewAPIGateway(substrateURL []string, identity substrate.Identity) (pkg.APIGateway, error) {
+	sub, err := substrate.NewManager(substrateURL...).Substrate()
+	if err != nil {
+		return nil, err
+	}
+	return &apiGateway{sub: sub, mu: sync.Mutex{}, identity: identity}, nil
+}
+
+func (g *apiGateway) CreateNode(node substrate.Node) (uint32, error) {
+	log.Debug().
+		Str("method", "CreateNode").
+		Uint32("twin id", uint32(node.TwinID)).
+		Uint32("farm id", uint32(node.FarmID)).
+		Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.CreateNode(g.identity, node)
+}
+
+func (g *apiGateway) CreateTwin(relay string, pk []byte) (uint32, error) {
+	log.Debug().Str("method", "CreateTwin").Str("relay", relay).Str("pk", hex.EncodeToString(pk)).Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.CreateTwin(g.identity, relay, pk)
+}
+
+func (g *apiGateway) EnsureAccount(activationURL string, termsAndConditionsLink string, termsAndConditionsHash string) (info substrate.AccountInfo, err error) {
+	log.Debug().
+		Str("method", "EnsureAccount").
+		Str("activation url", activationURL).
+		Str("terms and conditions link", termsAndConditionsLink).
+		Str("terms and conditions hash", termsAndConditionsHash).
+		Msg("method called")
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	return g.sub.EnsureAccount(g.identity, activationURL, termsAndConditionsLink, termsAndConditionsHash)
+}
+
+func (g *apiGateway) GetContract(id uint64) (substrate.Contract, pkg.SubstrateError) {
+	log.Trace().Str("method", "GetContract").Uint64("id", id).Msg("method called")
+	var SubstrateError pkg.SubstrateError
+	contract, err := g.sub.GetContract(id)
+	if contract == nil {
+		contract = &substrate.Contract{}
+	}
+	if errors.Is(err, substrate.ErrNotFound) {
+		SubstrateError.Err = err
+		SubstrateError.Code = pkg.CodeNotFound
+	} else if err != nil {
+		SubstrateError.Err = err
+		SubstrateError.Code = pkg.CodeGenericError
+	}
+	return *contract, SubstrateError
+}
```

Comment:
I suggest a different implementation approach for functions like these to make it easier for u as follows
```suggestion
func (g *apiGateway) GetContract(id uint64) (result substrate.Contract, serr pkg.SubstrateError) {
	log.Trace().Str("method", "GetContract").Uint64("id", id).Msg("method called")
	
	contract, err := g.sub.GetContract(id)
	serr.Err = err
	
	if errors.Is(err, substrate.ErrNotFound) {
		serr.Code = pkg.CodeNotFound
		return
	} else if err != nil {
		serr.Code = pkg.CodeGenericError
		return
	}
	// we sure that contract is not null here for sure
	return *contract, serr
}
```

Reply:
Also a good thing is if you can have an error builder function 
```go
serr := buildError(err)
```

So it fills up all the serr fields correctly (code, and whatever) for all known substate error this then can reduce the code above to 

```go
	contract, err := g.sub.GetContract(id)
	
         serr = buildErr(err)
         If err != nil { 
                 return 
         }

        return *contract, serr
```